lib/foreman/renderer/scope/macros/host_template.rb

```diff
@@ -121,6 +121,39 @@ def ks_console
             (@port && @baud) ? "console=ttyS#{@port},#{@baud}" : ''
           end
 
+          apipie :method, 'Generates commands for package(s) installation, supported OS families: Redhat, Debian & Suse' do
+            required :packages, String, desc: "Package(s) to install"
+            returns String, desc: 'Installation commands'
+            example "CentOS host: install_packages('pkg1') #=> yum -y install pkg1"
+            example "Fedora host: install_packages('pkg1 pkg2') #=> dnf -y install pkg1 pkg2"
+          end
+          def install_packages(packages)
+            raise UndefinedPackages.new if packages.empty?
```

Comment:
Why is it needed to raise an exception for this? Can't it just return nil or an empty string? This would make it easier to use and I don't seen downsides.

Reply:
Exception in macros are typically used to signal something unexpected happened and the exception message tells user what. I think the idea behind this was, if user calls the macro with empty list, it's probably not what they intended to do. However I agree with @ekohl that in this case, packages are mostly fed from the parameters which can be empty. Accepting (and ignoring) the empty list here would save users from having to add a condition in the template and checking for packages value presence.

While this is a general macro that can be also used without parameters, I think it's acceptable to not raise an error on empty string, it can return empty string in such case.